content/guides/weird_characters.adoc

```diff
@@ -0,0 +1,794 @@
+
+* https://github.com/bbatsov/clojure-style-guide#naming[Clojure Style Guide]
+
+== _ - Irrelevant var
+
+When you see this used as function arguments or similar, it is a common
+naming convention for vars or arguments you are not interested in using.
+That is you don't intend to use them, so you aren't really interested in
+thinking of a useful name for them.
+
+This is an example using the `add-watch` function that can be used to add
+callback style behaviour when atoms change value. Imagine, given an atom, we
+want to print the new value every time it changes
+[source,clojure]
+----
+(def value (atom 0))
+
+(add-watch value nil (fn [_ _ _ new-value]
+                       (println new-value))
+
+(reset! value 6)
+; prints 6
+(reset! value 9)
+; prints 9
+----
+`add-watch` takes four arguments, but in our case we only really care about the
+last argument - the new value of the atom.
+
+== #? - Standard Reader conditional
+
+Reader conditionals are designed to allow
+different dialects of Clojure to share common code. The standard reader
+conditional behaves similarly to a traditional `cond`. The syntax for usage
+is `#?` and looks like:
+[source,clojure]
+----
+#?(:clj  (Clojure expression)
+   :cljs (ClojureScript expression)
+   :cljr (Clojure CLR expression)
+   :default (fallthrough expression))
+----
+* <<xref/../reader_conditionals#,Reader conditonals>>
+
+== #@ - Splicing Reader conditional
+
+The syntax for a splicing reader conditional is `#?@`. It is used to splice
+lists into the containing form. So the Clojure reader would read this:
+[source,clojure]
+----
+(defn build-list []
+  (list #?@(:clj  [5 6 7 8]
+            :cljs [1 2 3 4])))
+----
+as this:
+[source,clojure]
+----
+(defn build-list []
+  (list 5 6 7 8))
+----
+* <<xref/../reader_conditionals#,Reader conditonals>>
+
+== #: Map Namespace Syntax
+
+Map namespace syntax was added in Clojure 1.9 and is used to specify a default
+namespace context for keys in the map using a `#:ns` prefix, where _ns_ is the
+name of a namespace and the prefix precedes teh opening brace `{` of the map.
+
+For example, the following map literal with namespace syntax:
+[source,clojure]
+----
+#:person{:first "Han"
+         :last "Solo"
+         :ship #:ship{:name "Millenium Falcon"
+                      :model "YT-1300f light freighter"}}
+----
+is read as:
+[source,clojure]
+----
+{:person/first "Han"
+ :person/last "Solo"
+ :person/ship {:ship/name "Millenium Falcon"
+               :ship/model "YT-1300f light freighter"}}
+----
+
+* <<xref/../../reference/reader#map_namespace_syntax,Reader>>
+
+== +#::+ Autoresolving Namespace Syntax
+
+`#::` can be used to auto-resolve namespaces with the same semantics as
+<<xref/../weird_characters#autoresolved_keys,autoresolved keywords.>>.
+
+* <<xref/../../reference/reader#map_namespace_syntax,Reader>>
+
+== #= Reader eval
+
+`#=` allows the reader to evaluate the following form.
+Examples
+[source,clojure]
+----
+#=123
```

Comment:
An arbitrary form can be evaluated during read time:

```
user=> (read-string "#=(+ 3 4)")
7
```

Not that the read-time evaluation can also cause side-effects:

```
user=> (read-string "#=(println :foo)")
:foo
nil
```

Consequently clojure.core/read-string is not safe to call with unverified user input. For a safe alternative, see clojure.edn/read-string. https://clojure.github.io/clojure/clojure.edn-api.html#clojure.edn/read-string

Note that `#=` is not an officially supported feature of the reader, so you shouldn't rely on its presence in future versions of Clojure.